# Matching and merging two datasets

Thank you for your interest in Finquest. To know more about your background and innovative drive in real-world data science/machine learning, we have curated two toy datasets for you to get your hands dirty.

## Goal

Your goal is to analyze these two datasets and think of as many as possible approaches to match rows in two datasets. This is an open question and your performance will be evaluated based on the following factors:

* Number of approaches proposed, implemented and evaluated.
* Ability to approach real-world problems with common data science tools and machine learning algorithms.
* Code quality of your implementation.

In general, we value both the broadness and depth of your takes on this challenge.


## Deadline

You have one week (7 days) to solve this challenge and send back your solutions as a *single* Jupyter Notebook (`.ipynb` format) to daoyuan.li@finquest.com.


## Notes

* Please work on this challenge by yourself and do not share it with any other persons.
* You own copyrights of your solutions.
* You are free to choose your programming language for this challenge, but Python 3.6+ is preferred if you join us later.
* It is OK to ask for clarifications to daoyuan.li@finquest.com.

We hope you enjoy this challenge, and we look forward to receiving your reply.

In [1]:
! wc -l *.tsv

  1826170 source1.tsv
  3531391 source2.tsv
  5357561 total


In [2]:
! head -n 10 *.tsv

==> source1.tsv <==
id	name	street_number	street_type	street_name	address_line2	postal_code	city
00542012000015	SOCIETE DES SUCRERIES DU MARQUENTERRE		RUE	DE LA FONTAINE		80120	RUE
00542012000031	SOCIETE DES SUCRERIES DU MARQUENTERRE		RTE	NATIONALE		62140	MARCONNELLE
00545011900028	LE GRAND CAFE ET GASSENDI			SAINT CHRISTOPHE	ROUTE DE MARSEILLE	04000	DIGNE-LES-BAINS
00552017600016	HERNAS CARTONNAGE	50	RUE	PASTEUR		80210	FEUQUIERES-EN-VIMEU
00552017600032	HERNAS CARTONNAGE			ZONE DES HAYETTES		80210	CHEPY
00552024200016	PROSPA	3	RUE	DU 43 RIC		80510	LONGPRE-LES-CORPS-SAINTS
00552032500019	ETABLISSEMENTS ADRIEN RIQUIER	12	RUE	HENRI BARBUSSE		80570	DARGNIES
00554155200013	ETABLISSEMENTS BONNET FRERES			HOTEL DU VERDON		04370	BEAUVEZER
00554155200039	ETABLISSEMENTS BONNET FRERES			PISTE DE L'AIGUILLE		04260	ALLOS

==> source2.tsv <==
address	website	id	name
226 r Menthon, 01380 SAINT CYR SUR MENTHON			Sandrine Berny
50 rte Bage, 01380 SAINT CYR SUR MENTHON		84122632700010	Sarl Landrix Frèr

### You may consider rows with the same `id` across different sources to be the same...

In [ ]:
722033800011

# Exploration 

In [1]:
import numpy as np
import pandas as pd
import os
from collections import Counter

ROOT = os.getcwd()

In [35]:
source2 = pd.read_csv(os.path.join(ROOT, 'source2.csv'))
source2

address  \
0               226 r Menthon, 01380 SAINT CYR SUR MENTHON   
1                 50 rte Bage, 01380 SAINT CYR SUR MENTHON   
2               120 r Menthon, 01380 SAINT CYR SUR MENTHON   
3        Moulin 28 chem RUILLAT, 01380 SAINT CYR SUR ME...   
4               350 rte Loeze, 01380 SAINT CYR SUR MENTHON   
...                                                    ...   
3530297                                                NaN   
3530298                       quart Guardère, 83150 BANDOL   
3530299                          5 r Jardins, 83150 BANDOL   
3530300                54 traverse Laboureur, 83150 BANDOL   
3530301                       9 r Tonneliers, 83150 BANDOL   

                                    website            id  \
0                                       NaN           NaN   
1                                       NaN  8.412263e+13   
2                                       NaN  7.913458e+13   
3                                       NaN           NaN   
4                                       NaN  4.341618e+13   
...                                     ...           ...   
3530297                                 NaN  8.497810e+13   
3530298  http://agence.carrefourlocation.fr           NaN   
3530299          http://aquabike-en-mer.com  7.951850e+13   
3530300     http://dolcefarnientebandol.com  5.392586e+13   
3530301                                 NaN  4.009196e+13   

                         name  
0              Sandrine Berny  
1         Sarl Landrix Frères  
2          Marion Kwasniewski  
3            CLAUDE FONDATION  
4           Manigand Stéphane  
...                       ...  
3530297              Jacoby O  
3530298    Carrefour Location  
3530299        Calvet Richard  
3530300      Dolce Far Niente  
3530301  Cordonnerie Nouvelle  

[3530302 rows x 4 columns]

In [31]:
source2 = source2.astype({'id':'Int64'})
source2

address  \
0               226 r Menthon, 01380 SAINT CYR SUR MENTHON   
1                 50 rte Bage, 01380 SAINT CYR SUR MENTHON   
2               120 r Menthon, 01380 SAINT CYR SUR MENTHON   
3        Moulin 28 chem RUILLAT, 01380 SAINT CYR SUR ME...   
4               350 rte Loeze, 01380 SAINT CYR SUR MENTHON   
...                                                    ...   
3530297                                                NaN   
3530298                       quart Guardère, 83150 BANDOL   
3530299                          5 r Jardins, 83150 BANDOL   
3530300                54 traverse Laboureur, 83150 BANDOL   
3530301                       9 r Tonneliers, 83150 BANDOL   

                                    website              id  \
0                                       NaN            <NA>   
1                                       NaN  84122632700010   
2                                       NaN  79134582000025   
3                                       NaN            <NA>   
4                                       NaN  43416176600010   
...                                     ...             ...   
3530297                                 NaN  84978102600018   
3530298  http://agence.carrefourlocation.fr            <NA>   
3530299          http://aquabike-en-mer.com  79518496900025   
3530300     http://dolcefarnientebandol.com  53925863200014   
3530301                                 NaN  40091957700025   

                         name  
0              Sandrine Berny  
1         Sarl Landrix Frères  
2          Marion Kwasniewski  
3            CLAUDE FONDATION  
4           Manigand Stéphane  
...                       ...  
3530297              Jacoby O  
3530298    Carrefour Location  
3530299        Calvet Richard  
3530300      Dolce Far Niente  
3530301  Cordonnerie Nouvelle  

[3530302 rows x 4 columns]

In [10]:
from recordlinkage.preprocessing import clean

In [11]:
#s1_s2_commonids = s1_s2_commonids.drop('Unnamed: 0', axis=1)
s1_s2_commonids['name_x'] = clean(s1_s2_commonids['name_x'], strip_accents='unicode')
s1_s2_commonids['name_y'] = clean(s1_s2_commonids['name_y'], strip_accents='unicode')
s1_s2_commonids['street_name'] = clean(s1_s2_commonids['street_name'], strip_accents='unicode')
s1_s2_commonids['street_type'] = clean(s1_s2_commonids['street_type'], strip_accents='unicode')
s1_s2_commonids['address_line2'] = clean(s1_s2_commonids['address_line2'], strip_accents='unicode')
s1_s2_commonids['address'] = clean(s1_s2_commonids['address'], strip_accents='unicode')
s1_s2_commonids['city'] = clean(s1_s2_commonids['city'], strip_accents='unicode')
s1_s2_commonids

/home/thabib/anaconda3/envs/nlp_siddata/lib/python3.8/site-packages/recordlinkage/preprocessing/cleaning.py:114: FutureWarning: The default value of regex will change from True to False in a future version.
  s = s.str.replace(r'(\[.*?\]|\(.*?\)|\{.*?\})', '')
/home/thabib/anaconda3/envs/nlp_siddata/lib/python3.8/site-packages/recordlinkage/preprocessing/cleaning.py:118: FutureWarning: The default value of regex will change from True to False in a future version.
  s = s.str.replace(replace_by_none, '')
/home/thabib/anaconda3/envs/nlp_siddata/lib/python3.8/site-packages/recordlinkage/preprocessing/cleaning.py:121: FutureWarning: The default value of regex will change from True to False in a future version.
  s = s.str.replace(replace_by_whitespace, ' ')
/home/thabib/anaconda3/envs/nlp_siddata/lib/python3.8/site-packages/recordlinkage/preprocessing/cleaning.py:124: FutureWarning: The default value of regex will change from True to False in a future version.
  s = s.str.replace(r'\s\s+',

id                                 name_x  street_number  \
0         542012000015  societe des sucreries du marquenterre            NaN   
1         542012000031  societe des sucreries du marquenterre            NaN   
2         552017600016                      hernas cartonnage           50.0   
3         552017600032                      hernas cartonnage            NaN   
4         552024200016                                 prospa            3.0   
...                ...                                    ...            ...   
781300  99889300200118        entreprise jean perraud et fils            NaN   
781301  99889300200142        entreprise jean perraud et fils           19.0   
781302  99889340800018                bijouterie dorey freres           27.0   
781303  99891270300019        materiel electrique longovicien            NaN   
781304  99892210800076                   etablissements laiti            NaN   

       street_type        street_name            address_line2  postal_code  \
0              rue     de la fontaine                      NaN        80120   
1              rte          nationale                      NaN        62140   
2              rue            pasteur                      NaN        80210   
3              NaN  zone des hayettes                      NaN        80210   
4              rue          du 43 ric                      NaN        80510   
...            ...                ...                      ...          ...   
781300         NaN             peuras        parc dactivite de        38210   
781301          av        victor hugo                      NaN        38800   
781302         rue        jean jaures                      NaN        42420   
781303         rue      du pulventeux                      NaN        54400   
781304         rue         du maillet  centre commercial geric        57100   

                            city                                    address  \
0                            rue                  1 r la fontaine 80120 rue   
1                    marconnelle            rte nationale 62140 marconnelle   
2            feuquieres en vimeu     50 r pasteur 80210 feuquieres en vimeu   
3                          chepy                    zi hayettes 80210 chepy   
4       longpre les corps saints  3 r 43 ric 80510 longpre les corps saints   
...                          ...                                        ...   
781300                   tullins                441 av peuras 38210 tullins   
781301          le pont de claix   19 av victor hugo 38800 le pont de claix   
781302                   lorette             27 r jean jaures 42420 lorette   
781303                    longwy                 zi pulventeux 54400 longwy   
781304                thionville      linkling 1 r maillet 57100 thionville   

                                website                              name_y  
0                                   NaN        sucreries du marquenterre sa  
1                                   NaN           sucreries du marquenterre  
2                  http://www.hernas.fr               hernas cartonnage sas  
3                  http://www.hernas.fr                   hernas cartonnage  
4                 http://www.prospa.fr/                              prospa  
...                                 ...                                 ...  
781300     http://www.groupeperraud.com                      groupe perraud  
781301     http://www.groupeperraud.com                      groupe perraud  
781302  http://www.bijouterie-dorey.com             bijouterie dorey freres  
781303                              NaN  mel paties mel materiel electrique  
781304                              NaN                    chaussures laiti  

[781305 rows x 11 columns]

In [26]:
cols = list(s1_s2_commonids.columns)

# cols[:-1] = 'name_s2'
cols = ['id',
 'name_s1',
 'street_number',
 'street_type',
 'street_name',
 'address_line2',
 'postal_code',
 'city',
 'address',
 'website',
 'name_s2']

s1_s2_commonids.columns = cols

In [32]:
source2.to_csv('./source2.csv', index=False)

In [33]:
source2.to_csv('./source2.tsv', sep='\t', index=False)

In [ ]:
# DELETE ABOVE

In [2]:
source1 = pd.read_csv(os.path.join(ROOT, 'source1.csv'))
source2 = pd.read_csv(os.path.join(ROOT, 'source2.csv'))

In [146]:
display(source1)
len(source1)

s1 = source1
# 1826169 rows in source 1 with all unique IDs

id                                   name  street_number  \
0          542012000015  SOCIETE DES SUCRERIES DU MARQUENTERRE            NaN   
1          542012000031  SOCIETE DES SUCRERIES DU MARQUENTERRE            NaN   
2          545011900028              LE GRAND CAFE ET GASSENDI            NaN   
3          552017600016                      HERNAS CARTONNAGE           50.0   
4          552017600032                      HERNAS CARTONNAGE            NaN   
...                 ...                                    ...            ...   
1826164  90771119600010                            BIONORMANDE            NaN   
1826165  90771120400012                              AHELA BAT           18.0   
1826166  90771121200015                          DECO BATIMENT            7.0   
1826167  90760906900014                    CHATEAUX MERVEILLES           15.0   
1826168  90760963000013                             CAP-ARMADE          140.0   

        street_type           street_name       address_line2  postal_code  \
0               RUE        DE LA FONTAINE                 NaN        80120   
1               RTE             NATIONALE                 NaN        62140   
2               NaN      SAINT CHRISTOPHE  ROUTE DE MARSEILLE         4000   
3               RUE               PASTEUR                 NaN        80210   
4               NaN     ZONE DES HAYETTES                 NaN        80210   
...             ...                   ...                 ...          ...   
1826164         RUE  DE LA GRANDE FLANDRE         LOT 1 - 7 A        76270   
1826165         RUE         SAINT CLEMENT                 NaN        93200   
1826166         RUE       AUGUSTE BLANQUI                 NaN        93200   
1826167         RUE              DE MILAN                 NaN        75009   
1826168         RUE        DE TOCQUEVILLE                 NaN        75017   

                        city  
0                        RUE  
1                MARCONNELLE  
2            DIGNE-LES-BAINS  
3        FEUQUIERES-EN-VIMEU  
4                      CHEPY  
...                      ...  
1826164   NEUFCHATEL-EN-BRAY  
1826165          SAINT-DENIS  
1826166          SAINT-DENIS  
1826167              PARIS 9  
1826168             PARIS 17  

[1826169 rows x 8 columns]

In [135]:
source1['postal_code'].isnull().sum()

0

In [4]:
#source2['id'] = pd.to_numeric(source2['id'], errors='coerce').astype(np.NaN) #source2.fillna(np.nan, inplace=True)
display(source2.head(20))
len(source2)

# 3530302 rows in source2 with 1258224 NaN IDs – pd.isna(source2['id']).sum()
# 3530302 - 1258224 = 2272078 rows with proper IDs

address  \
0          226 r Menthon, 01380 SAINT CYR SUR MENTHON   
1            50 rte Bage, 01380 SAINT CYR SUR MENTHON   
2          120 r Menthon, 01380 SAINT CYR SUR MENTHON   
3   Moulin 28 chem RUILLAT, 01380 SAINT CYR SUR ME...   
4          350 rte Loeze, 01380 SAINT CYR SUR MENTHON   
5            41 r Centre, 01380 SAINT CYR SUR MENTHON   
6   116 chem Cochatieres, 01380 SAINT CYR SUR MENTHON   
7       Za 25 chem Mayer, 01380 SAINT CYR SUR MENTHON   
8      420 chem Pavillon, 01380 SAINT CYR SUR MENTHON   
9               Chanelet, 01380 SAINT CYR SUR MENTHON   
10                26 r Pasteur, 02850 BARZY SUR MARNE   
11        19 pass Ecoles, 01380 SAINT CYR SUR MENTHON   
12         860 rte Mâcon, 01380 SAINT CYR SUR MENTHON   
13         226 r Menthon, 01380 SAINT CYR SUR MENTHON   
14           6 pl Dom Perignon, 02850 BARZY SUR MARNE   
15  612 lieu-dit Deschamps, 01380 SAINT CYR SUR ME...   
16  les Druillets 99 chem Cochatieres, 01380 SAINT...   
17          50 rte Loeze, 01380 SAINT CYR SUR MENTHON   
18        13 r St Leu Marcilly, 02850 BARZY SUR MARNE   
19      151 rte Chanelet, 01380 SAINT CYR SUR MENTHON   

                                          website            id  \
0                                             NaN           NaN   
1                                             NaN  8.412263e+13   
2                                             NaN  7.913458e+13   
3                                             NaN           NaN   
4                                             NaN  4.341618e+13   
5   http://www.infirmiere-frederique-thevenard.fr  4.885464e+13   
6                                             NaN  4.181119e+13   
7                                             NaN           NaN   
8                                             NaN  4.909120e+13   
9                                             NaN           NaN   
10                                            NaN           NaN   
11                                            NaN           NaN   
12                         http://www.peugeot.fr/  4.832794e+13   
13                                            NaN           NaN   
14                http://cocteau.site-solocal.com  4.207493e+13   
15            http://bruno-magnetiseur.weebly.com  4.405862e+13   
16                http://www.assurancemarechal.fr           NaN   
17                          http://www.capanel.fr  4.793428e+13   
18                                            NaN  5.046992e+13   
19                                            NaN  4.344290e+13   

                             name  
0                  Sandrine Berny  
1             Sarl Landrix Frères  
2              Marion Kwasniewski  
3                CLAUDE FONDATION  
4               Manigand Stéphane  
5            Frédérique Thévenard  
6                  Garnier Xavier  
7               Terre D'Alliances  
8                   Choquart Marc  
9                  Goujon Nicolas  
10                   Rabiat Alain  
11                         Mairie  
12              Garage Montbarbon  
13                 Isabelle Alban  
14  Champagne Jean Cocteau & Fils  
15                  Bruno Berthet  
16           Marechal Demars Anne  
17                   SARL Capanel  
18           Husson Petitbon Jean  
19             Des Chanelets GAEC

3530302

In [132]:
source2['name'].isnull().sum()

18

In [7]:
# Coverting float IDs to Int64 (pandas type)
s2 = source2.astype({'id': 'Int64'})
display(s2)

address  \
0               226 r Menthon, 01380 SAINT CYR SUR MENTHON   
1                 50 rte Bage, 01380 SAINT CYR SUR MENTHON   
2               120 r Menthon, 01380 SAINT CYR SUR MENTHON   
3        Moulin 28 chem RUILLAT, 01380 SAINT CYR SUR ME...   
4               350 rte Loeze, 01380 SAINT CYR SUR MENTHON   
...                                                    ...   
3530297                                                NaN   
3530298                       quart Guardère, 83150 BANDOL   
3530299                          5 r Jardins, 83150 BANDOL   
3530300                54 traverse Laboureur, 83150 BANDOL   
3530301                       9 r Tonneliers, 83150 BANDOL   

                                    website              id  \
0                                       NaN            <NA>   
1                                       NaN  84122632700010   
2                                       NaN  79134582000025   
3                                       NaN            <NA>   
4                                       NaN  43416176600010   
...                                     ...             ...   
3530297                                 NaN  84978102600018   
3530298  http://agence.carrefourlocation.fr            <NA>   
3530299          http://aquabike-en-mer.com  79518496900025   
3530300     http://dolcefarnientebandol.com  53925863200014   
3530301                                 NaN  40091957700025   

                         name  
0              Sandrine Berny  
1         Sarl Landrix Frères  
2          Marion Kwasniewski  
3            CLAUDE FONDATION  
4           Manigand Stéphane  
...                       ...  
3530297              Jacoby O  
3530298    Carrefour Location  
3530299        Calvet Richard  
3530300      Dolce Far Niente  
3530301  Cordonnerie Nouvelle  

[3530302 rows x 4 columns]

### Removing all entries with NaN IDs from source2 and also removing all duplicate rows

In [10]:
s2_nond = s2.dropna(subset=['id']).drop_duplicates()
display(s2_nond)

address  \
1                 50 rte Bage, 01380 SAINT CYR SUR MENTHON   
2               120 r Menthon, 01380 SAINT CYR SUR MENTHON   
4               350 rte Loeze, 01380 SAINT CYR SUR MENTHON   
5                 41 r Centre, 01380 SAINT CYR SUR MENTHON   
6        116 chem Cochatieres, 01380 SAINT CYR SUR MENTHON   
...                                                    ...   
3530296                         2 av Château, 83150 BANDOL   
3530297                                                NaN   
3530299                          5 r Jardins, 83150 BANDOL   
3530300                54 traverse Laboureur, 83150 BANDOL   
3530301                       9 r Tonneliers, 83150 BANDOL   

                                               website              id  \
1                                                  NaN  84122632700010   
2                                                  NaN  79134582000025   
4                                                  NaN  43416176600010   
5        http://www.infirmiere-frederique-thevenard.fr  48854644100012   
6                                                  NaN  41811188600028   
...                                                ...             ...   
3530296                                            NaN  30130663500097   
3530297                                            NaN  84978102600018   
3530299                     http://aquabike-en-mer.com  79518496900025   
3530300                http://dolcefarnientebandol.com  53925863200014   
3530301                                            NaN  40091957700025   

                         name  
1         Sarl Landrix Frères  
2          Marion Kwasniewski  
4           Manigand Stéphane  
5        Frédérique Thévenard  
6              Garnier Xavier  
...                       ...  
3530296  Carlot Jean-François  
3530297              Jacoby O  
3530299        Calvet Richard  
3530300      Dolce Far Niente  
3530301  Cordonnerie Nouvelle  

[2271362 rows x 4 columns]

In [13]:
len(np.unique(s2_nond['id']))

2270350

__There are only 2270350 unique IDs in source2 and the remaining 1012 (=2271362 - 2270350) IDs must, therefore, be repititions__

In [53]:
s2nond_uids, s2nond_uids_counts = np.unique(np.array(s2_nond['id']), return_counts=True)
duplicates = s2nond_uids[s2nond_uids_counts > 1]
display(duplicates)
display(len(duplicates))

Counter(s2nond_uids_counts)

array([6480115200089, 7692068500039, 13002792300049, ..., 95750393100561,
       99785860000023, 99882350414085], dtype=object)

1011

Counter({1: 2269339, 2: 1010, 3: 1})

_There are 1010 unique IDs that are repeated twice and 1 unique ID is repeated thrice._

In [54]:
# IDs repeated twice
display(s2nond_uids[np.where(s2nond_uids_counts==2)])
# IDs repeated thrice
display(s2nond_uids[np.where(s2nond_uids_counts==3)])

array([6480115200089, 7692068500039, 13002792300049, ..., 95750393100561,
       99785860000023, 99882350414085], dtype=object)

array([38990291700019], dtype=object)

#### How many repeated rows are there in source2 with NAN IDs?

In [110]:
#s2_nod = s2.drop_duplicates()
s2_nanid = s2[~s2.isin(s2_nond)].dropna(how='all')#.drop_duplicates()
s2_nod_nanid = s2_nanid.drop_duplicates()

#del s2_nod
display(s2_nanid)
display(s2_nod_nanid)

address  \
0               226 r Menthon, 01380 SAINT CYR SUR MENTHON   
3        Moulin 28 chem RUILLAT, 01380 SAINT CYR SUR ME...   
7            Za 25 chem Mayer, 01380 SAINT CYR SUR MENTHON   
9                    Chanelet, 01380 SAINT CYR SUR MENTHON   
10                     26 r Pasteur, 02850 BARZY SUR MARNE   
...                                                    ...   
3530277           39 Ter r Doct Louis Marçon, 83150 BANDOL   
3530278                       66 r Lamartine, 83150 BANDOL   
3530287        Les Cotx St Mar r Jean Jaurès, 83150 BANDOL   
3530288                  7 r Vincent Allègre, 83150 BANDOL   
3530298                       quart Guardère, 83150 BANDOL   

                                       website    id  \
0                                          NaN  <NA>   
3                                          NaN  <NA>   
7                                          NaN  <NA>   
9                                          NaN  <NA>   
10                                         NaN  <NA>   
...                                        ...   ...   
3530277                                    NaN  <NA>   
3530278                                    NaN  <NA>   
3530287                                    NaN  <NA>   
3530288  http://heleneguermeur-psychologue.com  <NA>   
3530298     http://agence.carrefourlocation.fr  <NA>   

                                                      name  
0                                           Sandrine Berny  
3                                         CLAUDE FONDATION  
7                                        Terre D'Alliances  
9                                           Goujon Nicolas  
10                                            Rabiat Alain  
...                                                    ...  
3530277  Compagnie Mediterraneenne Des Multicoques De C...  
3530278                                  Chaux Pierre-Yves  
3530287                               Corbeau Marie-Pierre  
3530288                                    Guermeur Hélène  
3530298                                 Carrefour Location  

[1258932 rows x 4 columns]

address  \
0               226 r Menthon, 01380 SAINT CYR SUR MENTHON   
3        Moulin 28 chem RUILLAT, 01380 SAINT CYR SUR ME...   
7            Za 25 chem Mayer, 01380 SAINT CYR SUR MENTHON   
9                    Chanelet, 01380 SAINT CYR SUR MENTHON   
10                     26 r Pasteur, 02850 BARZY SUR MARNE   
...                                                    ...   
3530277           39 Ter r Doct Louis Marçon, 83150 BANDOL   
3530278                       66 r Lamartine, 83150 BANDOL   
3530287        Les Cotx St Mar r Jean Jaurès, 83150 BANDOL   
3530288                  7 r Vincent Allègre, 83150 BANDOL   
3530298                       quart Guardère, 83150 BANDOL   

                                       website    id  \
0                                          NaN  <NA>   
3                                          NaN  <NA>   
7                                          NaN  <NA>   
9                                          NaN  <NA>   
10                                         NaN  <NA>   
...                                        ...   ...   
3530277                                    NaN  <NA>   
3530278                                    NaN  <NA>   
3530287                                    NaN  <NA>   
3530288  http://heleneguermeur-psychologue.com  <NA>   
3530298     http://agence.carrefourlocation.fr  <NA>   

                                                      name  
0                                           Sandrine Berny  
3                                         CLAUDE FONDATION  
7                                        Terre D'Alliances  
9                                           Goujon Nicolas  
10                                            Rabiat Alain  
...                                                    ...  
3530277  Compagnie Mediterraneenne Des Multicoques De C...  
3530278                                  Chaux Pierre-Yves  
3530287                               Corbeau Marie-Pierre  
3530288                                    Guermeur Hélène  
3530298                                 Carrefour Location  

[1257964 rows x 4 columns]

In [83]:
1257964 + 2271362

3529326

In [84]:
s2_ = s2.drop_duplicates()
display(s2_)

address  \
0               226 r Menthon, 01380 SAINT CYR SUR MENTHON   
1                 50 rte Bage, 01380 SAINT CYR SUR MENTHON   
2               120 r Menthon, 01380 SAINT CYR SUR MENTHON   
3        Moulin 28 chem RUILLAT, 01380 SAINT CYR SUR ME...   
4               350 rte Loeze, 01380 SAINT CYR SUR MENTHON   
...                                                    ...   
3530297                                                NaN   
3530298                       quart Guardère, 83150 BANDOL   
3530299                          5 r Jardins, 83150 BANDOL   
3530300                54 traverse Laboureur, 83150 BANDOL   
3530301                       9 r Tonneliers, 83150 BANDOL   

                                    website              id  \
0                                       NaN            <NA>   
1                                       NaN  84122632700010   
2                                       NaN  79134582000025   
3                                       NaN            <NA>   
4                                       NaN  43416176600010   
...                                     ...             ...   
3530297                                 NaN  84978102600018   
3530298  http://agence.carrefourlocation.fr            <NA>   
3530299          http://aquabike-en-mer.com  79518496900025   
3530300     http://dolcefarnientebandol.com  53925863200014   
3530301                                 NaN  40091957700025   

                         name  
0              Sandrine Berny  
1         Sarl Landrix Frères  
2          Marion Kwasniewski  
3            CLAUDE FONDATION  
4           Manigand Stéphane  
...                       ...  
3530297              Jacoby O  
3530298    Carrefour Location  
3530299        Calvet Richard  
3530300      Dolce Far Niente  
3530301  Cordonnerie Nouvelle  

[3528618 rows x 4 columns]

In [107]:
other_ids_s2 = []
for idx, i in enumerate(s2['id']):
    if type(i) is np.int64 or type(i) is pd._libs.missing.NAType:
        continue
    else:
        other_ids_s2.append(i)

In [108]:
other_ids_s2

[]

### Checking if the entries are the same for IDs that are repeated (it is given above that they can be considered the same)

In [57]:
s2_nond[s2_nond['id'] == 99785860000023]

address  \
2255547  Rte de Remalard lieu-dit Tuilot, 61400 MORTAGN...   
2255597  Rte de Remalard lieu-dit Tuilot, 61400 MORTAGN...   

                                      website              id  \
2255547  http://www.intermarche.com/home.html  99785860000023   
2255597                                   NaN  99785860000023   

                                name  
2255547  Intermarché station-service  
2255597                  Intermarché

In [58]:
s2_nond[s2_nond['id'] == 6480115200089]

address  \
655852  Parc Du Banian 75 montée St Menet, 13011 MARSE...   
655899                75 montée St Menet, 13011 MARSEILLE   

                         website             id  \
655852  http://www.fidecompta.fr  6480115200089   
655899                       NaN  6480115200089   

                                  name  
655852                      Fidecompta  
655899  Fiduciaire Expertise Comptable

In [56]:
s2_nond[s2_nond['id'] == 38990291700019]

address                                 website  \
346816  av Bamberg, 12000 RODEZ  https://location.intermarche.com/site/   
503674  av Bamberg, 12000 RODEZ        https://location.intermarche.com   
566976  av Bamberg, 12000 RODEZ    http://www.intermarche.com/home.html   

                    id                        name  
346816  38990291700019  Intermarché location Rodez  
503674  38990291700019  Intermarché location Rodez  
566976  38990291700019           Intermarché Super

#### Constructing a dictionary with counter for each unique IDs number of occurences

In [16]:
dict_ids_counts = Counter(np.array(s2_nond['id']))


# Put IDs with 2 or more occurences in a list
s2nond_rep_ids = []
for k,v in dict_ids_counts.items():
    if v == 2 or v==3:
        s2nond_rep_ids.append(k)

### Checking how many of these repeating IDs from 's2_nond' are in 's1'

In [17]:
s2nond_rep_ids_in_s1, _ , s1_rep_idx = np.intersect1d(s2nond_rep_ids, s1['id'], return_indices=True)

In [18]:
display(s2nond_rep_ids_in_s1)
len(s1_rep_idx)

array([ 6480115200089,  7692068500039, 19762093300014, 20005562200084,
       20005684400521, 21050036900017, 21100117700057, 21130077700177,
       21130102300126, 21170415000187, 21250370000032, 21250453400042,
       21260148800043, 21260165200010, 21260170200039, 21260252800136,
       21270375500166, 21270562800023, 21280005600013, 21280051000027,
       21280358900028, 21280380300023, 21280387800025, 21290233200098,
       21290262100011, 21300008600024, 21310104100027, 21310375700091,
       21310396300020, 21310424300034, 21310495300020, 21310587700012,
       21320117100021, 21320160100027, 21330003100035, 21330029600026,
       21330153400029, 21330357100029, 21360182600059, 21380006300017,
       21440143200064, 21490223100148, 21510185800029, 21510602200241,
       21510614700022, 21590158800016, 21740133000030, 21760157400114,
       21760157400239, 21780015000096, 21780123200083, 21780123200133,
       21780380800062, 21830086100034, 21860093000018, 22720002900055,
      

421

In [80]:
display(s1[s1['id']==38990291700019])
display(s2_nond[s2_nond['id']==38990291700019])

id    name  street_number street_type street_name  \
426935  38990291700019  JANELI            NaN          AV  DE BAMBERG   

       address_line2  postal_code   city  
426935           NaN        12000  RODEZ

address                                 website  \
346816  av Bamberg, 12000 RODEZ  https://location.intermarche.com/site/   
503674  av Bamberg, 12000 RODEZ        https://location.intermarche.com   
566976  av Bamberg, 12000 RODEZ    http://www.intermarche.com/home.html   

                    id                        name  
346816  38990291700019  Intermarché location Rodez  
503674  38990291700019  Intermarché location Rodez  
566976  38990291700019           Intermarché Super

In [90]:
id_ = 40087862500013
display(s1[s1['id']==id_])
display(s2_nond[s2_nond['id']==id_])

id                                          name  \
481691  40087862500013  CENTRE DE CONTROLE TECHNIQUE DE LANN SEVELIN   

        street_number street_type             street_name       address_line2  \
481691            NaN         RUE  JEAN BAPTISTE MARTENOT  ZI DE LANN SEVELIN   

        postal_code    city  
481691        56850  CAUDAN

address website  \
1986022  Zone Industrielle De Lann Sevelin r Jean Bapti...     NaN   
1986538         350 r Jean Baptiste Martenot, 56850 CAUDAN     NaN   

                     id                                     name  
1986022  40087862500013   CENTRE CONTROLE TECHNIQUE LANN SEVELIN  
1986538  40087862500013  Contrôle Technique Lann Sévelin Kerpont

In [93]:
s2_nond.loc[1986022]['address']

'Zone Industrielle De Lann Sevelin r Jean Baptiste Martenot, 56850 CAUDAN'

In [58]:
id_ = 84238156800018
display(s1[s1['id']==id_])
display(s2_nond[s2_nond['id']==id_])

id      name  street_number street_type   street_name  \
1641863  84238156800018  SARL JYS           53.0         RUE  CESAR LENOIR   

        address_line2  postal_code             city  
1641863           NaN        50340  SIOUVILLE-HAGUE

address  \
1885389  53 r Cesar Lenoir, 50340 SIOUVILLE HAGUE   
1885421  53 r Cesar Lenoir, 50340 SIOUVILLE HAGUE   

                                                   website              id  \
1885389                                                NaN  84238156800018   
1885421  http://fr-fr.facebook.com/pages/category/Ice-C...  84238156800018   

                name  
1885389     SARL JYS  
1885421  Croquez Moi

In [59]:
id_ = 75147510400012
display(s1[s1['id']==id_])
display(s2_nond[s2_nond['id']==id_])

id              name  street_number street_type  \
1159164  75147510400012  OPTIQUE MEDICALE           52.0          AV   

             street_name address_line2  postal_code       city  
1159164  GEOFFROY PERRET           NaN        30210  REMOULINS

address  \
973460  52 av Geoffroy Perret, 30210 REMOULINS   
973884  52 av Geoffroy Perret, 30210 REMOULINS   

                                        website              id  \
973460                                      NaN  75147510400012   
973884  http://optiquemedicale.site-solocal.com  75147510400012   

                                         name  
973460   Optique Medicale Melle Chamand Celin  
973884  Optique Médicale - Vision et audition

### Checking how many IDs from 's2_nond' are in 's1'

In [138]:
s2nond_ids_in_s1, s1_idx , s2nond_idx = np.intersect1d(s1['id'], s2_nond['id'], return_indices=True)

In [139]:
display(s2nond_ids_in_s1)
len(s1_idx)

array([542012000015, 542012000031, 552017600016, ..., 99889340800018,
       99891270300019, 99892210800076], dtype=object)

780883

#### Creating a dataframe of common IDs

In [137]:
s1_s2_commonids = pd.merge(s1, s2_nond, on='id').dropna(subset=['id'])
display(s1_s2_commonids)

id                                 name_x  street_number  \
0         542012000015  SOCIETE DES SUCRERIES DU MARQUENTERRE            NaN   
1         542012000031  SOCIETE DES SUCRERIES DU MARQUENTERRE            NaN   
2         552017600016                      HERNAS CARTONNAGE           50.0   
3         552017600032                      HERNAS CARTONNAGE            NaN   
4         552024200016                                 PROSPA            3.0   
...                ...                                    ...            ...   
781300  99889300200118        ENTREPRISE JEAN PERRAUD ET FILS            NaN   
781301  99889300200142        ENTREPRISE JEAN PERRAUD ET FILS           19.0   
781302  99889340800018                BIJOUTERIE DOREY FRERES           27.0   
781303  99891270300019        MATERIEL ELECTRIQUE LONGOVICIEN            NaN   
781304  99892210800076                   ETABLISSEMENTS LAITI            NaN   

       street_type        street_name            address_line2  postal_code  \
0              RUE     DE LA FONTAINE                      NaN        80120   
1              RTE          NATIONALE                      NaN        62140   
2              RUE            PASTEUR                      NaN        80210   
3              NaN  ZONE DES HAYETTES                      NaN        80210   
4              RUE          DU 43 RIC                      NaN        80510   
...            ...                ...                      ...          ...   
781300         NaN             PEURAS       PARC D'ACTIVITE DE        38210   
781301          AV        VICTOR HUGO                      NaN        38800   
781302         RUE        JEAN JAURES                      NaN        42420   
781303         RUE      DU PULVENTEUX                      NaN        54400   
781304         RUE         DU MAILLET  CENTRE COMMERCIAL GERIC        57100   

                            city                                     address  \
0                            RUE                  1 r la Fontaine, 80120 RUE   
1                    MARCONNELLE            rte Nationale, 62140 MARCONNELLE   
2            FEUQUIERES-EN-VIMEU     50 r Pasteur, 80210 FEUQUIÈRES EN VIMEU   
3                          CHEPY                    zi Hayettes, 80210 CHÉPY   
4       LONGPRE-LES-CORPS-SAINTS  3 r 43 Ric, 80510 LONGPRÉ LES CORPS SAINTS   
...                          ...                                         ...   
781300                   TULLINS                441 av Peuras, 38210 TULLINS   
781301          LE PONT-DE-CLAIX   19 av Victor Hugo, 38800 LE PONT DE CLAIX   
781302                   LORETTE             27 r Jean Jaurès, 42420 LORETTE   
781303                    LONGWY                 zi Pulventeux, 54400 LONGWY   
781304                THIONVILLE      Linkling 1 r Maillet, 57100 THIONVILLE   

                                website  \
0                                   NaN   
1                                   NaN   
2                  http://www.hernas.fr   
3                  http://www.hernas.fr   
4                 http://www.prospa.fr/   
...                                 ...   
781300     http://www.groupeperraud.com   
781301     http://www.groupeperraud.com   
781302  http://www.bijouterie-dorey.com   
781303                              NaN   
781304                              NaN   

                                        name_y  
0                 Sucreries du Marquenterre SA  
1                    Sucreries du Marquenterre  
2                        Hernas Cartonnage SAS  
3                            Hernas Cartonnage  
4                                       Prospa  
...                                        ...  
781300                          Groupe Perraud  
781301                          Groupe Perraud  
781302                 Bijouterie Dorey Frères  
781303  Mel - Patiès M.E.L Matériel Electrique  
781304                        Chaussures Laiti  

[781305 rows x 11 columns]

In [143]:
781305-780883

422

In [140]:
s1_s2_commonids['id'].isnull().sum()

0

In [142]:
len(np.unique(s1_s2_commonids['id']))

780883

### Counting NANs in columns of source1 and 2

In [35]:
s1['city'].isnull().sum()

0

### Counting full address (no NANs in any address field) in source1

In [38]:
s1_fulladdress = s1.dropna(subset=['street_number', 'street_type', 'street_name', 'address_line2'])
s1_fulladdress

id                                               name  \
19         568014500017                    PROMENS MONTOIR DE BRETAGNE SAS   
29         572078400148                            ETABLISSEMENTS DECAYEUX   
30         572078400155                            ETABLISSEMENTS DECAYEUX   
42         578096000109  SOCIETE IMMOBILIERE TOURISTIQUE ET HOTELIERE D...   
46         582037800031                              ENTREPRISE DEMOUSELLE   
...                 ...                                                ...   
1826064  90771049500017                                              GIPSY   
1826065  90771050300018                                   MEYYARTIST PARIS   
1826085  90771083400017                                     LARC BATISSEUR   
1826150  90771098200014                               LAPOLEONE PUBLISHING   
1826160  90771111300023                   CABINET DU DOCTEUR ARNAUD ALLUIN   

         street_number street_type         street_name  \
19                73.0         RUE       HENRI GAUTIER   
29                 1.0         IMP    NICEPHORE NIEPCE   
30                 7.0         RUE       DES SARCELLES   
42                 5.0         ESP      FRANCOIS ANDRE   
46               140.0         RUE    DU CHATEAU D EAU   
...                ...         ...                 ...   
1826064           10.0         RUE          DE TREVISE   
1826065           76.0          AV  DES CHAMPS ELYSEES   
1826085           15.0          AV   FERNAND CHATELAIN   
1826150            4.0         RUE             GALILEE   
1826160          199.0         RUE     DE LA RIANDERIE   

                             address_line2  postal_code                 city  
19                         "LA PROVIDENCE"        44550  MONTOIR-DE-BRETAGNE  
29                    ZA CHARLES DE GAULLE        93290   TREMBLAY-EN-FRANCE  
30                       ZONE INDUSTRIELLE        80100            ABBEVILLE  
42          SUR PLAGE FACE HOTEL HERMITAGE        44500   LA BAULE-ESCOUBLAC  
46                       ZONE INDUSTRIELLE        80100            ABBEVILLE  
...                                    ...          ...                  ...  
1826064                              10-12        75009              PARIS 9  
1826065                              76-78        75008              PARIS 8  
1826085            BATIMENT SUNSET PORTE 4        95610               ERAGNY  
1826150                                4-6        75116             PARIS 16  
1826160  CLINIQUE DU SPORT ET DE CHIRURGIE        59700     MARCQ-EN-BAROEUL  

[270690 rows x 8 columns]

# Writing to disk

In [112]:
s2_nond.to_csv('./source2_nonanids_nod.tsv', sep='\t')

In [113]:
s2_nond_ = pd.read_csv(os.path.join(ROOT, 'source2_nonanids_nod.tsv'), sep='\t')
display(s2_nond_)

Unnamed: 0                                            address  \
0                 1           50 rte Bage, 01380 SAINT CYR SUR MENTHON   
1                 2         120 r Menthon, 01380 SAINT CYR SUR MENTHON   
2                 4         350 rte Loeze, 01380 SAINT CYR SUR MENTHON   
3                 5           41 r Centre, 01380 SAINT CYR SUR MENTHON   
4                 6  116 chem Cochatieres, 01380 SAINT CYR SUR MENTHON   
...             ...                                                ...   
2271357     3530296                         2 av Château, 83150 BANDOL   
2271358     3530297                                                NaN   
2271359     3530299                          5 r Jardins, 83150 BANDOL   
2271360     3530300                54 traverse Laboureur, 83150 BANDOL   
2271361     3530301                       9 r Tonneliers, 83150 BANDOL   

                                               website              id  \
0                                                  NaN  84122632700010   
1                                                  NaN  79134582000025   
2                                                  NaN  43416176600010   
3        http://www.infirmiere-frederique-thevenard.fr  48854644100012   
4                                                  NaN  41811188600028   
...                                                ...             ...   
2271357                                            NaN  30130663500097   
2271358                                            NaN  84978102600018   
2271359                     http://aquabike-en-mer.com  79518496900025   
2271360                http://dolcefarnientebandol.com  53925863200014   
2271361                                            NaN  40091957700025   

                         name  
0         Sarl Landrix Frères  
1          Marion Kwasniewski  
2           Manigand Stéphane  
3        Frédérique Thévenard  
4              Garnier Xavier  
...                       ...  
2271357  Carlot Jean-François  
2271358              Jacoby O  
2271359        Calvet Richard  
2271360      Dolce Far Niente  
2271361  Cordonnerie Nouvelle  

[2271362 rows x 5 columns]

In [119]:
s2_nod_nanid.to_csv('./source2_nanids_nod.tsv', sep='\t')

In [120]:
s2_nod_nanid_ = pd.read_csv(os.path.join(ROOT, 'source2_nanids_nod.tsv'), sep='\t')
display(s2_nod_nanid_)

Unnamed: 0                                            address  \
0                 0         226 r Menthon, 01380 SAINT CYR SUR MENTHON   
1                 3  Moulin 28 chem RUILLAT, 01380 SAINT CYR SUR ME...   
2                 7      Za 25 chem Mayer, 01380 SAINT CYR SUR MENTHON   
3                 9              Chanelet, 01380 SAINT CYR SUR MENTHON   
4                10                26 r Pasteur, 02850 BARZY SUR MARNE   
...             ...                                                ...   
1257959     3530277           39 Ter r Doct Louis Marçon, 83150 BANDOL   
1257960     3530278                       66 r Lamartine, 83150 BANDOL   
1257961     3530287        Les Cotx St Mar r Jean Jaurès, 83150 BANDOL   
1257962     3530288                  7 r Vincent Allègre, 83150 BANDOL   
1257963     3530298                       quart Guardère, 83150 BANDOL   

                                       website  id  \
0                                          NaN NaN   
1                                          NaN NaN   
2                                          NaN NaN   
3                                          NaN NaN   
4                                          NaN NaN   
...                                        ...  ..   
1257959                                    NaN NaN   
1257960                                    NaN NaN   
1257961                                    NaN NaN   
1257962  http://heleneguermeur-psychologue.com NaN   
1257963     http://agence.carrefourlocation.fr NaN   

                                                      name  
0                                           Sandrine Berny  
1                                         CLAUDE FONDATION  
2                                        Terre D'Alliances  
3                                           Goujon Nicolas  
4                                             Rabiat Alain  
...                                                    ...  
1257959  Compagnie Mediterraneenne Des Multicoques De C...  
1257960                                  Chaux Pierre-Yves  
1257961                               Corbeau Marie-Pierre  
1257962                                    Guermeur Hélène  
1257963                                 Carrefour Location  

[1257964 rows x 5 columns]

In [144]:
s1_s2_commonids.to_csv('./s1_s2_commonids.tsv', sep='\t')

In [145]:
s1_s2_commonids_ = pd.read_csv(os.path.join(ROOT, 's1_s2_commonids.tsv'), sep='\t')
display(s1_s2_commonids_)

Unnamed: 0              id                                 name_x  \
0                0    542012000015  SOCIETE DES SUCRERIES DU MARQUENTERRE   
1                1    542012000031  SOCIETE DES SUCRERIES DU MARQUENTERRE   
2                2    552017600016                      HERNAS CARTONNAGE   
3                3    552017600032                      HERNAS CARTONNAGE   
4                4    552024200016                                 PROSPA   
...            ...             ...                                    ...   
781300      781300  99889300200118        ENTREPRISE JEAN PERRAUD ET FILS   
781301      781301  99889300200142        ENTREPRISE JEAN PERRAUD ET FILS   
781302      781302  99889340800018                BIJOUTERIE DOREY FRERES   
781303      781303  99891270300019        MATERIEL ELECTRIQUE LONGOVICIEN   
781304      781304  99892210800076                   ETABLISSEMENTS LAITI   

        street_number street_type        street_name            address_line2  \
0                 NaN         RUE     DE LA FONTAINE                      NaN   
1                 NaN         RTE          NATIONALE                      NaN   
2                50.0         RUE            PASTEUR                      NaN   
3                 NaN         NaN  ZONE DES HAYETTES                      NaN   
4                 3.0         RUE          DU 43 RIC                      NaN   
...               ...         ...                ...                      ...   
781300            NaN         NaN             PEURAS       PARC D'ACTIVITE DE   
781301           19.0          AV        VICTOR HUGO                      NaN   
781302           27.0         RUE        JEAN JAURES                      NaN   
781303            NaN         RUE      DU PULVENTEUX                      NaN   
781304            NaN         RUE         DU MAILLET  CENTRE COMMERCIAL GERIC   

        postal_code                      city  \
0             80120                       RUE   
1             62140               MARCONNELLE   
2             80210       FEUQUIERES-EN-VIMEU   
3             80210                     CHEPY   
4             80510  LONGPRE-LES-CORPS-SAINTS   
...             ...                       ...   
781300        38210                   TULLINS   
781301        38800          LE PONT-DE-CLAIX   
781302        42420                   LORETTE   
781303        54400                    LONGWY   
781304        57100                THIONVILLE   

                                           address  \
0                       1 r la Fontaine, 80120 RUE   
1                 rte Nationale, 62140 MARCONNELLE   
2          50 r Pasteur, 80210 FEUQUIÈRES EN VIMEU   
3                         zi Hayettes, 80210 CHÉPY   
4       3 r 43 Ric, 80510 LONGPRÉ LES CORPS SAINTS   
...                                            ...   
781300                441 av Peuras, 38210 TULLINS   
781301   19 av Victor Hugo, 38800 LE PONT DE CLAIX   
781302             27 r Jean Jaurès, 42420 LORETTE   
781303                 zi Pulventeux, 54400 LONGWY   
781304      Linkling 1 r Maillet, 57100 THIONVILLE   

                                website  \
0                                   NaN   
1                                   NaN   
2                  http://www.hernas.fr   
3                  http://www.hernas.fr   
4                 http://www.prospa.fr/   
...                                 ...   
781300     http://www.groupeperraud.com   
781301     http://www.groupeperraud.com   
781302  http://www.bijouterie-dorey.com   
781303                              NaN   
781304                              NaN   

                                        name_y  
0                 Sucreries du Marquenterre SA  
1                    Sucreries du Marquenterre  
2                        Hernas Cartonnage SAS  
3                            Hernas Cartonnage  
4                                       Prospa  
...                                        ...  
781300      

# EDA Summary

__For Source1__

1. Total rows = 1826169
2. Unique rows = 1826169
3. Rows with non-NAN IDs = 1826169
4. Rows with unique non-NAN IDs = 1826169


5. NANs in 'name' = 2
6. Full addresses (no NANs in any of the address fields) = 270690




__For Source2__
1. Total rows = 3530302
2. Unique rows = 3528618 
3. Rows with non-NAN IDs = 2272078 
4. Rows with NAN IDs = 1257964
5. Rows left after removing duplicate rows and NAN IDs = 2271362. 
6. Rows with unique non-NAN IDs = 2270350.
7. Number of rows with repetitions of IDs = 2271362-2270350 = 1012 _{1010 IDs repeating twice, 1 ID repeating thrice}_


(For s2_nond)

8. NANs in 'address' = 10448, with 754 common with source1 (on IDs)
9. NANs in 'name' = 6
10. NANs in 'website' = 1320267



__Source1 + Source2__

1. Number of IDs common to both = 780883
2. Hence, there are 1045286 rows in source1 (_not in source2_) which have a unique IDs but missing values of different address fields and name etc.
3. Also, there are 1490479 (=2271362-780883) rows in source2 (_not in source1_) which have (non-NAN) IDs (1489467 (=2270350-780883) unique) but missing values in 'address' and 'name' fields.


# Strategies

<font color='blue'>

    
1. Remove the similar (repetitive IDs) entries from source2 using NER.

    
2. Parse 'address' field of source2 into corresponding fields from source1.

    
3. Match 780883 rows in source1 and souce2 with common IDs, using information in the fields from each table to complete missing entries in the other table (if any)
    
 
</font>